### DISCUSSION TITLE
How to list all functions and fields in userdata

### DISCUSSION THREAD

**Discussion by lammermann** [mlua-rs/mlua#481] [Q&A]
Hello guys,
I try to build a REPL with mlua and rustyline. 

When it comes to tab-completion I check if the current prompt could be completed by using the method `for_each` from the tables. For example given the current prompt `foo.b` it searches the lua globals for `foo` and then searches all fields in `foo` and filters to only those beginning with `b` then returning `bar` and `baz` as possible completions.
This works fine for tables but I could not find something similar for userdata ( I found the method `pairs` in the `UserDataMetatable` but it does not seem to return the name of the functions of the userdata).

Is there any advice how to get this information from userdata?

Thanks in advance for any hints

**Comment by cheesycod** [comment]
The best way to do this is to honestly have each userdata implement an __iter metatable (or whatever your Lua uses for this)
